### ISSUE TITLE
t.w._flatten._flattenElement should support t.p.urlpath.URLPath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6238]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6238|
|Type|enhancement|
|Created|2012-12-31 17:22:34Z|

I think the right thing to do here is to turn it into a string.

Attachments:

* [6238_v1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d8c/d8c41f8fa5403ecd97e1937281361fbe338c68cb/8a6ef766ac614fff08cec6b80d99d8268d1e8909.diff) (2728 bytes) - added by borko on 2013-01-04 00:17:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6238 6238
type__enhancement enhancement
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__ 
branch_author__borko borko
status__closed closed
resolution__wontfix wontfix
component__web web
keywords__web_templates web templates
time__1356974554000000 1356974554000000
changetime__1357338286000000 1357338286000000
version__None None
owner__ 
cc__jknight cc__borko
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Hi borko.  Thanks for working on this.

Sticking to the high-level, it's fairly important that the URL be properly escaped in the output of `flatten`.  This patch so far makes no attempt to apply the proper escaping.  Proper escaping here is made more difficult, I think, by the fact that `URLPath` doesn't bother to document what the result of using `str()` on it will be, nor whether it its various attributes store encoded or decoded data.

I think perhaps it's a bad idea to try to write any code that uses `URLPath`, due to the shortcomings of `URLPath`.  A better course would be to work on improving `URLPath` (or replacing it) to the point where it's possible to be confident in uses of it (in other words, I don't think it is possible to have a satisfactory resolution to this ticket at this time).

At the very minimum, the tests for this feature need to include coverage for the various special characters a URL can include, and the two different flattener cases - `inAttribute = True` and `inAttribute = False`.

**Comment by twisted-trac** [comment]
|<img alt="borko's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|borko set status to `closed`|
|-|-|

Ok. So I opt for creating new ticket for changing URLPath in a way it should be done with precise comment, setting this ticket as blocked and writing short note about this fact including a link to the ticket that must be done before. Thx.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="borko's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|borko commented|
|-|-|

so what's the final status of this ticket? are we closing it without any changes to URLPath ?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Just to be clear about my changes, I was simply (dumbly) reacting to the fact that this was closed as "fixed".  If it hasn't actually been fixed but it's just waiting to be reopened, 'wontfix' is a more appropriate status.  I haven't had time to review it to see if further action is warranted :).
